docs: add ceph-exporter design doc (#11360)
rook/rook: design/ceph/ceph-exporter.md

# Ceph exporter

> ### Goals

Reviewer: nit: add a blank line before and after all titles
```suggestion
### Goals

```

Use Ceph exporter as the new source of metrics based on performance counters coming from every Ceph daemon.

## Use case
Ceph exporter was created to avoid performance problems at scale obtaining metrics in clusters with high number of daemons (mainly OSDs).
The Prometheus manager module was not enough performant to obtain the performance counters from all the daemons along the whole Ceph cluster, so the solution for avoid this problem in big clusters is to have a new daemon collecting all the performance counters from all the daemons in each host, and expose them as standard prometheus metrics at /metrics endpoint for prometheus to scrape each of those endpoints.
### Why do we need this daemon in Rook

> Although probably the Rook clusters usually do not have a huge number of daemons, and therefore, probably the main problem solved by the Ceph Exporter(scalability) is not going to be present, with the Ceph exporter, the source of metrics changes in the Ceph clusters.

Reviewer: Is the central mgr module still going to exist? Or the new ceph exporter daemon will be required? If both are options, seems like we might want to have an option for smaller clusters to still report to the mgr, and larger clusters to report to the ceph exporter. But if the only supported option will be the exporter, then we will always need to start it.

We do have many large clusters in rook upstream, so let's not assume rook doesn't benefit from the change. 
```suggestion
With the Ceph exporter, the source of metrics changes in the Ceph clusters.
```

Author: Yes, we still need prometheus module exporter for cluster-wide metrics. Well, at least in latest quincy we still haven't removed the perf counters metrics support from prometheus module, but we have ceph-exporter too to do this job. 
After this release we're removing perf counters from module and only exporter will take care of this. But we've discussed to introduce a module option for prometheus module to enable/disable support for fetching perf counters metrics (`default: false`)

Author: Additionally, `ceph-exporter` should be deployed by default in rook cluster (recommended). The module option will be useful in case if exporter pods failed to start or any issues with container not running then user still can have option to get metrics from module.

After introducing Ceph Exporter in Ceph cluster we are going to have two main sources of metrics:

> - Prometheus manager module: It is responsible for exposing all metrics other than ceph daemons performance counters.

Reviewer: If the exporter is not running, will the ceph daemons still report to this module?

Author: > If the exporter is not running, will the ceph daemons still report to this module?

https://github.com/rook/rook/pull/11360#discussion_r1035593705

- Ceph exporter: It is responsible for exposing only ceph daemons performance counters as prometheus metrics.

So if we do not deploy the Ceph exporter we are not going to have access to the daemon performance counters (in form of metrics) coming from the different Ceph daemons.

## Prerequisites
- Ceph daemon socket files must have ceph user permission

> - hostPath for ceph daemon socket dirs

Reviewer: ```suggestion
- A hostPath is available for ceph daemon socket dirs
```

Approach / Implementation details

> ## Proposal details

Reviewer: With this title, seems like you can delete line 20.
```suggestion
## Implementation details
```

Ceph-exporter is a daemon for collecting perf counters of ceph daemons from admin socket files and exposing them to /metrics endpoint.
So we will need the following things:

> - Deploy it as a daemonset in all the k8s cluster nodes where a Ceph daemon is running

Reviewer: Instead of a daemonset, let's approach it similar to the crash collector pod. Rook creates a crash collector deployment on every node where it detects a ceph daemon. The same controller could also be updated to start an exporter deployment on the same nodes.

> - nodeAffinity

Reviewer: ?

Author: nodeAffinity to make sure exporter pods gets deployed only on ceph nodes.

- hostNetwork: true

> - RestartPolicy : Always

Reviewer: No need for the host network and restart policy settings for the design doc. Or if it's an important detail, you can change it to sentence form.

- hostPath for ceph daemon socket dirs
- Volumes & VolumeMounts for ceph pods & containers for daemon socket files and also ceph config file (/var/lib/rook/rook-ceph/rook-ceph.config)
- Add daemonSocketDir to ChownCephDataDirsInitContainer()
- Provide a /metrics endpoint: It needs a new K8s service.
- Needs to be scraped by prometheus: We need to use a Prometheus Service Monitor which select the Ceph exporter service

> #### Example

Reviewer: It would help to show an example of the services and the service monitor yaml that will be created to help illustrate the behavior (and they are much shorter than a pod spec). For example:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: ceph-exporter-<node>
  namespace: <namespace>
  labels:
    app: rook-ceph-exporter
    node: <node>
spec:
  internalTrafficPolicy: Cluster
  ports:
  - name: http-metrics
    port: 9283
    protocol: TCP
    targetPort: 9283
  selector:
    app: rook-ceph-exporter
    node: <nodeName>
  sessionAffinity: None
  type: ClusterIP
```

And the ServiceMonitor would have the `rook-ceph-exporter` selector (without the `node` selector), so hopefully the single service monitor would collect metrics from all the nodes' services.
```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: rook-ceph-exporter
  namespace: <namespace>
spec:
  namespaceSelector:
    matchNames:
      - <namespace>
  selector:
    matchLabels:
      app: rook-ceph-exporter
  endpoints:
  - port: http-metrics
    path: /metrics
    interval: 5s
```

Below is an example Ceph-exporter CRD, `ceph-exporter.yaml`

> ```yaml

Reviewer: I'd suggest no need for the export spec in the design doc. The properties listed above are sufficient.

# Ceph exporter

### Goals

Use Ceph exporter as the new source of metrics based on performance counters coming from every Ceph daemon.

## Use case

Ceph exporter was created to avoid performance problems at scale obtaining metrics in clusters with high number of daemons (mainly OSDs).
The Prometheus manager module was not enough performant to obtain the performance counters from all the daemons along the whole Ceph cluster, so the solution for avoid this problem in big clusters is to have a new daemon collecting all the performance counters from all the daemons in each host, and expose them as standard prometheus metrics at /metrics endpoint for prometheus to scrape each of those endpoints.

### Why do we need this daemon in Rook

With the Ceph exporter, the source of metrics changes in the Ceph clusters.
There will be two main sources of metrics:
- Prometheus manager module: It is responsible for exposing all metrics other than ceph daemons performance counters.
- Ceph exporter: It is responsible for exposing only ceph daemons performance counters as prometheus metrics.

> Ceph exporter will be deployed in Rook when the `monitoring.enabled: true` setting is applied in the CephCluster CR and the Ceph version later than 17.2.5.

Reviewer: If it's not merged yet, how about we include a link to the ceph change so we can track it here
```suggestion
Ceph exporter will be deployed in Rook when the `monitoring.enabled: true` setting is applied in the CephCluster CR and the Ceph version planned starting with 17.2.6. The Ceph change is PR <link>
```

Author: PR is merged and has been backported to quincy as well. It''s just I wasn't sure which next version to mention, so I wrote later than 17.2.5

Reviewer: Ok sounds good if it's merged and backported to quincy

## Prerequisites

- Ceph daemon socket files must have ceph user permission
- A hostPath is available for ceph daemon socket dirs

## Implementation details

Ceph-exporter is a daemon for collecting perf counters of ceph daemons from admin socket files and exposing them to /metrics endpoint.
So we will need the following things:
- Deploy it similar to crash collector pod

> - hostPath for ceph daemon socket dirs

Reviewer: Can we declare a directory under the `dataDirHostPath`? For example: `$dataDirHostPath/exporter`

Author: Yes, for hostPath I used [`var/lib/ceph/sock-dir`](https://github.com/rook/rook/pull/11317/files#diff-8c0c2ac670977110df3027659350706f78b42f5132b2cc765a8e9172b6cfdca4R35)

Reviewer: How about mentioning it in the doc? Note that `/var/lib/ceph` is only mounted in the container. The host path is (by default) under `/var/lib/rook`
```suggestion
- hostPath for ceph daemon socket dirs under `$dataDirHostPath/exporter`
```

- Volumes & VolumeMounts for ceph pods & containers for daemon socket files and also ceph config file (/var/lib/rook/rook-ceph/rook-ceph.config)
- Add daemonSocketDir to ChownCephDataDirsInitContainer()
- Provide a /metrics endpoint: It needs a new K8s service.
- Needs to be scraped by prometheus: We need to use a Prometheus Service Monitor which select the Ceph exporter service

> #### Exporter Service

Reviewer: higher level title
```suggestion
### Exporter Service
```

> 

Reviewer: Could you add a sentence that clarifies this will be created for each node where an exporter is running?

`ceph-exporter-service.yaml`
```yaml
apiVersion: v1
kind: Service
metadata:
  name: ceph-exporter-<node>
  namespace: <namespace>
  labels:
    app: rook-ceph-exporter
    node: <node>
spec:
  internalTrafficPolicy: Cluster
  ports:
  - name: http-metrics
    port: 9926
    protocol: TCP
    targetPort: 9926
  selector:
    app: rook-ceph-exporter
    node: <nodeName>
  sessionAffinity: None
  type: ClusterIP
```

> Exporter ServiceMonitor

Reviewer: How about a title
```suggestion
### Exporter ServiceMonitor
```

> ```yaml

Reviewer: How about a sentence describing that this is a single service monitor that will match all the exporter services.

# Ceph exporter

### Goals

Use Ceph exporter as the new source of metrics based on performance counters coming from every Ceph daemon.

## Use case

Ceph exporter was created to avoid performance problems at scale obtaining metrics in clusters with high number of daemons (mainly OSDs).
The Prometheus manager module was not enough performant to obtain the performance counters from all the daemons along the whole Ceph cluster, so the solution for avoid this problem in big clusters is to have a new daemon collecting all the performance counters from all the daemons in each host, and expose them as standard prometheus metrics at /metrics endpoint for prometheus to scrape each of those endpoints.

### Why do we need this daemon in Rook

With the Ceph exporter, the source of metrics changes in the Ceph clusters.
There will be two main sources of metrics:
- Prometheus manager module: It is responsible for exposing all metrics other than ceph daemons performance counters.
- Ceph exporter: It is responsible for exposing only ceph daemons performance counters as prometheus metrics.

> Ceph exporter will be deployed by default in rook ceph cluster. But in case if pod fails to start or some issues with starting the container, it is still possible to fetch perf counters metrics from prometheus module by setting module option `fetch_perf_counters_metrics` to true.\

Reviewer: If you want to mention it's still possible to enable the module setting, let's mention it separately to be clear that rook will not enable it, and we don't see the need for it since we can behave differently depending on the ceph version.
```suggestion
Ceph exporter will be deployed in Rook when the `monitoring.enabled: true` setting is applied in the CephCluster CR, and the Ceph version is at least 17.2.x **(insert expected version)**`
```

Author: Ok makes sense! and by separately you mean somewhere in rook ceph documentation right?

  `ceph config set mgr mgr/prometheus/fetch_perf_counters_metrics true`

## Prerequisites

- Ceph daemon socket files must have ceph user permission
- A hostPath is available for ceph daemon socket dirs

## Implementation details

Ceph-exporter is a daemon for collecting perf counters of ceph daemons from admin socket files and exposing them to /metrics endpoint.
So we will need the following things:
- Deploy it as a daemonset in all the k8s cluster nodes where a Ceph daemon is running
- nodeAffinity to make sure exporter pods gets deployed only on ceph nodes.
- hostPath for ceph daemon socket dirs
- Volumes & VolumeMounts for ceph pods & containers for daemon socket files and also ceph config file (/var/lib/rook/rook-ceph/rook-ceph.config)
- Add daemonSocketDir to ChownCephDataDirsInitContainer()
- Provide a /metrics endpoint: It needs a new K8s service.
- Needs to be scraped by prometheus: We need to use a Prometheus Service Monitor which select the Ceph exporter service

> #### Example of services and service monitor yamls for `ceph-exporter`

Reviewer: ```suggestion
### Exporter Service
```

`ceph-exporter-service.yaml`
```yaml
apiVersion: v1
kind: Service
metadata:
  name: ceph-exporter-<node>
  namespace: <namespace>
  labels:
    app: rook-ceph-exporter
    node: <node>
spec:
  internalTrafficPolicy: Cluster
  ports:
  - name: http-metrics
    port: 9926
    protocol: TCP
    targetPort: 9926
  selector:
    app: rook-ceph-exporter
    node: <nodeName>
  sessionAffinity: None
  type: ClusterIP
```

> ServiceMonitor for ceph-exporter

Reviewer: ```suggestion
### Exporter ServiceMonitor
```
